src/codegen/src/main/scala/PySparkWrapper.scala

```diff
@@ -238,39 +258,37 @@ abstract class PySparkWrapper(entryPoint: PipelineStage,
       val pname = param.name
       val docType = getPythonizedDataType(param.getClass.getSimpleName)
       paramGettersAndSetters +=
-        setTemplate.mkString("\n").format(StringUtils.capitalize(pname), pname, docType, ScopeDepth * 2,
-          getParamExplanation(param))
-      if(isComplexType(param.getClass.getSimpleName)){
+        setTemplate(StringUtils.capitalize(pname), pname,
+          paramDocTemplate(getParamExplanation(param), docType, ScopeDepth * 3))
+      if (isComplexType(param.getClass.getSimpleName)) {
         paramDefinitionsAndDefaults +=
-          defineComplexParamsTemplate.format(
-            pname, getParamExplanation(param),
+          defineComplexParamsTemplate(pname, getParamExplanation(param),
             s"""generateTypeConverter("$pname", self._cache, complexTypeConverter)""")
         paramGettersAndSetters +=
-          getComplexTemplate.mkString("\n").format(StringUtils.capitalize(pname), pname, param.getClass.getSimpleName)
+          getComplexTemplate(StringUtils.capitalize(pname), pname, docType, getParamExplanation(param))
         paramDocList +=
-          paramDocTemplate.format(pname, getParamExplanation(param), ScopeDepth * 2, param.getClass.getSimpleName)
+          paramDocTemplate(getParamExplanation(param), docType, ScopeDepth * 3)
         classParamDocList +=
-          paramDocTemplate.format(pname, getParamExplanation(param), ScopeDepth, param.getClass.getSimpleName)
-      }
-      else{
+          paramDocTemplate(getParamExplanation(param), docType, ScopeDepth * 2)
+      } else {
         paramDefinitionsAndDefaults +=
-          defineParamsTemplate.format(pname, getParamExplanation(param))
+          s"""        self.$pname = Param(self, \"$pname\", \"${getParamExplanation(param)}\")"""
         paramGettersAndSetters +=
-          getTemplate.mkString("\n").format(StringUtils.capitalize(pname), pname, docType)
+          getTemplate(StringUtils.capitalize(pname), pname, docType, getParamExplanation(param))
         paramDocList +=
-          paramDocTemplate.format(pname, getParamExplanation(param), ScopeDepth * 2, docType)
+          paramDocTemplate(getParamExplanation(param), docType, ScopeDepth * 3)
         classParamDocList +=
-          paramDocTemplate.format(pname, getParamExplanation(param), ScopeDepth, param.getClass.getSimpleName)
+          paramDocTemplate(getParamExplanation(param), docType, ScopeDepth * 2)
       }
 
       val (pyParamDefault, autogenSuffix) = getParamDefault(param)
       paramsAndDefaults += pname + "=" + pyParamDefault
 
-      if (pyParamDefault != "None") {
-        paramDefinitionsAndDefaults += setParamDefaultTemplate.format(pname, pyParamDefault)
-      } else if (autogenSuffix != null) {
-        paramDefinitionsAndDefaults += setParamDefaultWithGuidTemplate.format(pname, autogenSuffix)
-      }
+      if (pyParamDefault != "None")
+        paramDefinitionsAndDefaults += s"""        self._setDefault($pname=$pyParamDefault)"""
```

Comment:
The fixed number of spaces concerns me a bit especially since they're significant. Is it possible to explicitly specify the number of spaces?